### DISCUSSION TITLE
Katex and Mathjax in Nextjs

### DISCUSSION THREAD

**Discussion by edward1127** [vercel/next.js#54852] [Help]
### Summary

      
Given https://codesandbox.io/s/katex-mhchem-wl7q64 and https://stackblitz.com/edit/nextjs-xvqkxb?file=pages/index.js clearly `rehype-katex` clearly can access the mhchem context.
https://stackblitz.com/edit/stackblitz-starters-ueq3rk?file=src%2FApp.tsx Shows that everything works fine in React.
With the smaller reproducible examples it's increasingly looking like it works up to the point `next` it added to the picture.
Which implies something `next` is doing is likely the root cause.

Combined with https://github.com/remarkjs/remark-math/issues/80 it leads me to believe something about Next SSR messes with Katex and Mathjax.
Remark and MDX just happen to also be in the dependency tree, but seem to be unrelated to the issues being seen.

_Originally posted by @ChristianMurphy in https://github.com/remarkjs/remark-math/issues/81#issuecomment-1441197972_

I try to use CSR with Next https://stackblitz.com/edit/nextjs-ftepsg?file=pages%2Findex.js, and it's still not working. I believe somehow nextjs cannot get the context of `import "katex/dist/contrib/mhchem";`

### Example
vanilla js
https://codesandbox.io/s/katex-mhchem-wl7q64
react
https://stackblitz.com/edit/stackblitz-starters-ueq3rk?file=src%2FApp.tsx  
next
https://stackblitz.com/edit/nextjs-xvqkxb?file=pages/index.js
next csr
https://stackblitz.com/edit/nextjs-ftepsg?file=pages%2Findex.js

**Comment by AAC7** [comment]
### NEXT JS SOLUTION

```
import katex from "katex";
import "katex/dist/katex.min.css";
import { useEffect, useRef } from "react";
import MathRenderer from "./MathRenderer";
import "katex/contrib/mhchem";

interface ChemistryEquationProps {
  text: string;
}

const ChemistryEquation = ({ text }: ChemistryEquationProps) => {
  const containerRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const renderChem = async () => {
      // mhchem registers itself with katex internally
      const html = katex.renderToString(`\\ce{${text}}`, {
        throwOnError: true,
      });

      if (containerRef.current) {
        containerRef.current.innerHTML = html;
      }
    };

    renderChem();
  }, [text]);

  return <span ref={containerRef} />;
};

export default ChemistryEquation;

``` 

usage 
Again make sure using in client

    <ChemistryEquation text="Acid reaction: \ce{x 2Na 2(NH4)HPO4 ->[catalysit][neeche] (NaPO3)_x + x NH3 ^ + x H2O}" />


try this make sure its client component not ssr
